plugins/modules/ovirt_disk.py

```diff
@@ -420,118 +429,175 @@ def create_transfer_connection(module, transfer, context, connect_timeout=10, re
-            otypes.ImageTransferPhase.TRANSFERRING,
-            otypes.ImageTransferPhase.FINALIZING_SUCCESS,
-        ]:
-            time.sleep(module.params['poll_interval'])
+    while True:
+        time.sleep(1)
+        try:
             transfer = transfer_service.get()
-        if transfer.phase in [
-            otypes.ImageTransferPhase.UNKNOWN,
-            otypes.ImageTransferPhase.FINISHED_FAILURE,
-            otypes.ImageTransferPhase.FINALIZING_FAILURE,
-            otypes.ImageTransferPhase.CANCELLED,
-        ]:
-            raise Exception(
-                "Error occurred while uploading image. The transfer is in %s" % transfer.phase
-            )
-        if not module.params.get('logical_unit'):
-            disks_service = connection.system_service().disks_service()
-            wait(
-                service=disks_service.service(module.params['id']),
-                condition=lambda d: d.status == otypes.DiskStatus.OK,
-                wait=module.params['wait'],
-                timeout=module.params['timeout'],
-            )
+        except sdk.NotFoundError:
+            # The system has removed the disk and the transfer.
+            raise RuntimeError("Transfer {0} was removed".format(transfer.id))
+
+        if transfer.phase == otypes.ImageTransferPhase.FINISHED_FAILURE:
+            # The system will remove the disk and the transfer soon.
+            raise RuntimeError("Transfer {0} has failed".format(transfer.id))
+
+        if transfer.phase == otypes.ImageTransferPhase.CANCELLED_SYSTEM:
+            # The system will remove the disk and the transfer soon.
+            raise RuntimeError("Transfer {0} was cancelled by system".format(transfer.id))
+
+        if transfer.phase == otypes.ImageTransferPhase.PAUSED_SYSTEM:
+            transfer_service.cancel()
+            raise RuntimeError(
+                "Transfer {0} was paused by system".format(transfer.id))
+
+        if transfer.phase == otypes.ImageTransferPhase.TRANSFERRING:
+            break
+
+        if transfer.phase != otypes.ImageTransferPhase.INITIALIZING:
+            transfer_service.cancel()
+            raise RuntimeError(
+                "Unexpected transfer {0} phase {1}"
+                .format(transfer.id, transfer.phase))
+
+        if time.time() > start + module.params.get('timeout'):
+            transfer_service.cancel()
+            raise RuntimeError(
+                "Timed out waiting for transfer {0}".format(transfer.id))
+
+    hosts_service = connection.system_service().hosts_service()
+    host_service = hosts_service.host_service(transfer.host.id)
+    transfer.host = host_service.get()
+    return transfer
+
+
+def cancel_transfer(connection, transfer):
+    transfer_service = (connection.system_service()
+                        .image_transfers_service()
+                        .image_transfer_service(transfer.id))
+    transfer_service.cancel()
+
+
+def finalize_transfer(connection, module, transfer):
```

Comment:
It will be cleaner to pass "transfer.id" instead.
You don't actually use the "transfer" object and almost immediately override it.